### PR TITLE
Participant UUID field

### PR DESCRIPTION
**Pull request by iambibhas**

### CODE REVIEW

File: funnel/templates/scan_badge.html.jinja2 [hunk 1]
```diff
@@ -72,7 +72,7 @@
     $(function() {
 
       var eventConfig = {
-        checkinApiUrl: "{{ url_for('checkin_puk', profile=project.profile.name, project=project.name, name=event.name, puk='puk') }}",
+        checkinApiUrl: "{{ url_for('EventParticipantCheckinView_checkin_puk', profile=project.profile.name, project=project.name, event=event.name, puk='puk') }}",
```
Comment (jace) [inline funnel/templates/scan_badge.html.jinja2]: This should be `event.url_for(...)`

File: funnel/views/participant.py [hunk 2]
```diff
@@ -48,64 +51,83 @@ def participant_badge_data(participants, project):
...
 
 @route('/<profile>/<project>/participants')
 class ProjectParticipantView(ProjectViewMixin, UrlForView, ModelView):
     __decorators__ = [legacy_redirect]
 
     @route('json')
+    @render_with(json=True)
     @lastuser.requires_login
     @requires_permission('view')
     def participants_json(self):
```
Comment (jace) [inline funnel/views/participant.py]: This view should be dropped. It is a security risk. It was added for speedy badge scanning in the native app.

File: funnel/views/participant.py [hunk 3]
```diff
@@ -48,64 +51,83 @@ def participant_badge_data(participants, project):
...
-        return jsonify(
-            participants=[
+        return {
+            'participants': [
                 participant_data(participant, self.obj.id)
```
Comment (jace) [inline funnel/views/participant.py]: This view is also leaking database primary key for the project.

File: funnel/views/participant.py [hunk 4]
```diff
@@ -130,190 +152,181 @@ class FunnelProjectParticipantView(ProjectParticipantView):
...
-    (Profile, {'name': 'profile'}, 'g.profile'),
-    ((Project, ProjectRedirect), {'name': 'project', 'profile': 'profile'}, 'project'),
-    (Participant, {'id': 'participant_id'}, 'participant'),
-    permission='checkin_event',
-)
-def participant_label_badge(profile, project, participant):
-    return render_template(
-        'label_badge.html.jinja2', badges=participant_badge_data([participant], project)
-    )
+@route('/<profile>/<project>/participant/<suuid>')
```
Comment (jace) [inline funnel/views/participant.py]: Make the URL parameter `<participant>`

File: funnel/views/participant.py [hunk 5]
```diff
@@ -130,190 +152,181 @@ class FunnelProjectParticipantView(ProjectParticipantView):
...
+class ParticipantView(UrlForView, ModelView):
+    __decorators__ = [legacy_redirect, lastuser.requires_login]
 
+    model = Participant
+    route_model_map = {
+        'profile': 'project.profile.name',
+        'project': 'project.name',
+        'suuid': 'suuid',
```
Comment (jace) [inline funnel/views/participant.py]: LHS = `'participant'`

File: funnel/views/participant.py [hunk 6]
```diff
@@ -130,190 +152,181 @@ class FunnelProjectParticipantView(ProjectParticipantView):
...
-        if request.is_xhr:
-            return jsonify(
-                status=True, participant_ids=participant_ids, checked_in=checked_in
+    def loader(self, profile, project, suuid):
+        participant = (
+            self.model.query.join(Project, Profile)
+            .filter(
+                Profile.name == profile,
+                Project.name == project,
+                Participant.suuid == suuid,
```
Comment (jace) [inline funnel/views/participant.py]: RHS = `participant`

File: funnel/views/participant.py [hunk 7]
```diff
@@ -130,190 +152,181 @@ class FunnelProjectParticipantView(ProjectParticipantView):
...
-        checked_in = getbool(request.form.get('checkin'))
-        participant_ids = request.form.getlist('pid')
-        for participant_id in participant_ids:
-            attendee = Attendee.get(event, participant_id)
-            attendee.checked_in = checked_in
-        db.session.commit()
-        if request.is_xhr:
-            return jsonify(
-                status=True, participant_ids=participant_ids, checked_in=checked_in
+    def loader(self, profile, project, suuid):
```
Comment (jace) [inline funnel/views/participant.py]: Accept `participant` instead of `suuid`

File: funnel/views/participant.py [hunk 8]
```diff
@@ -130,190 +152,181 @@ class FunnelProjectParticipantView(ProjectParticipantView):
...
+                Participant.suuid == suuid,
             )
-    return redirect(event.url_for('view'), code=303)
+            .first_or_404()
+        )
+        return participant
 
+    def after_loader(self):
+        g.profile = self.obj.project.profile
+        super(ParticipantView, self).after_loader()
```
Comment (jace) [inline funnel/views/participant.py]: This should be a `return super...` as `after_loader` can optionally return a response.

File: funnel/templates/scan_badge.html.jinja2 (hunk 1 above)
Comment (iambibhas) [reply funnel/templates/scan_badge.html.jinja2]: This endpoint is a bit weirdly handled on the front end. Moving it to `event.url_for()` will require redoing front end as well. I'll check with vidya if it's possible. I thought we'll come and revisit this `puk` issue later and for now do this.

File: funnel/views/mixins.py [hunk 9]
```diff
@@ -198,6 +199,29 @@ def loader(self, profile, project, venue, room):
...
+        'name': 'name',
+    }
+
+    def loader(self, profile, project, name):
+        event = (
+            self.model.query.join(Project, Profile)
+            .filter(
+                Profile.name == profile, Project.name == project, Event.name == name
+            )
+            .first_or_404()
```
Comment (jace) [inline funnel/views/mixins.py:216]: `one_or_404()`

**Pull request merged**